In [2]:
# Import Library
import pandas as pd

In [3]:
#Import the files to Google Colab
url = 'https://raw.githubusercontent.com/rc-dbe/bigdatacertification/master/dataset/churn_trasnsformed_new.csv'
df_csv = pd.read_csv(url, sep=',',)

# Show 10 first Row
df_csv.head()

Unnamed: 0  gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  \
0           0         0         1                0                1   
1           1         1         0                0                1   
2           2         1         0                0                1   
3           3         1         0                0                1   
4           4         0         1                0                1   

   Partner_0  Partner_1  Dependents_0  Dependents_1  tenure  ...  Contract_1  \
0          0          1             0             1       1  ...           0   
1          1          0             0             1      34  ...           1   
2          1          0             0             1       2  ...           0   
3          1          0             0             1      45  ...           1   
4          1          0             0             1       2  ...           0   

   Contract_2  PaperlessBilling_0  PaperlessBilling_1  PaymentMethod_0  \
0           1                   0                   1                0   
1           0                   1                   0                0   
2           1                   0                   1                0   
3           0                   1                   0                0   
4           1                   0                   1                0   

   PaymentMethod_1  PaymentMethod_2  MonthlyCharges  TotalCharges  Churn  
0                0                1            2985        2985.0      0  
1                1                0            5695       18895.0      0  
2                1                0            5385       10815.0      1  
3                1                1             423      184075.0      0  
4                0                1             707       15165.0      1  

[5 rows x 47 columns]

In [4]:
# Remove "Unnamed:O" Coloumn
df = df_csv.drop("Unnamed: 0", axis=1)
df.head()

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  Partner_1  \
0         0         1                0                1          0          1   
1         1         0                0                1          1          0   
2         1         0                0                1          1          0   
3         1         0                0                1          1          0   
4         0         1                0                1          1          0   

   Dependents_0  Dependents_1  tenure  PhoneService_0  ...  Contract_1  \
0             0             1       1               0  ...           0   
1             0             1      34               1  ...           1   
2             0             1       2               1  ...           0   
3             0             1      45               0  ...           1   
4             0             1       2               1  ...           0   

   Contract_2  PaperlessBilling_0  PaperlessBilling_1  PaymentMethod_0  \
0           1                   0                   1                0   
1           0                   1                   0                0   
2           1                   0                   1                0   
3           0                   1                   0                0   
4           1                   0                   1                0   

   PaymentMethod_1  PaymentMethod_2  MonthlyCharges  TotalCharges  Churn  
0                0                1            2985        2985.0      0  
1                1                0            5695       18895.0      0  
2                1                0            5385       10815.0      1  
3                1                1             423      184075.0      0  
4                0                1             707       15165.0      1  

[5 rows x 46 columns]

In [5]:
# Check the Data Information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 46 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   gender_0            7043 non-null   int64  
 1   gender_1            7043 non-null   int64  
 2   SeniorCitizen_0     7043 non-null   int64  
 3   SeniorCitizen_1     7043 non-null   int64  
 4   Partner_0           7043 non-null   int64  
 5   Partner_1           7043 non-null   int64  
 6   Dependents_0        7043 non-null   int64  
 7   Dependents_1        7043 non-null   int64  
 8   tenure              7043 non-null   int64  
 9   PhoneService_0      7043 non-null   int64  
 10  PhoneService_1      7043 non-null   int64  
 11  MultipleLines_0     7043 non-null   int64  
 12  MultipleLines_1     7043 non-null   int64  
 13  MultipleLines_2     7043 non-null   int64  
 14  InternetService_0   7043 non-null   int64  
 15  InternetService_1   7043 non-null   int64  
 16  Intern

In [6]:
#Import MinMax Scaler
from sklearn.preprocessing import MinMaxScaler

# initialize min-max scaler
mm_scaler = MinMaxScaler()
column_names = df.columns.tolist()
column_names.remove('Churn')

# Transform all attributes
df[column_names] = mm_scaler.fit_transform(df[column_names])
df.sort_index(inplace=True)
df.head()

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  Partner_1  \
0       0.0       1.0              0.0              1.0        0.0        1.0   
1       1.0       0.0              0.0              1.0        1.0        0.0   
2       1.0       0.0              0.0              1.0        1.0        0.0   
3       1.0       0.0              0.0              1.0        1.0        0.0   
4       0.0       1.0              0.0              1.0        1.0        0.0   

   Dependents_0  Dependents_1    tenure  PhoneService_0  ...  Contract_1  \
0           0.0           1.0  0.013889             0.0  ...         0.0   
1           0.0           1.0  0.472222             1.0  ...         1.0   
2           0.0           1.0  0.027778             1.0  ...         0.0   
3           0.0           1.0  0.625000             0.0  ...         1.0   
4           0.0           1.0  0.027778             1.0  ...         0.0   

   Contract_2  PaperlessBilling_0  PaperlessBilling_1  PaymentMethod_0  \
0         1.0                 0.0                 1.0              0.0   
1         0.0                 1.0                 0.0              0.0   
2         1.0                 0.0                 1.0              0.0   
3         0.0                 1.0                 0.0              0.0   
4         1.0                 0.0                 1.0              0.0   

   PaymentMethod_1  PaymentMethod_2  MonthlyCharges  TotalCharges  Churn  
0              0.0              1.0        0.250169      0.003420      0  
1              1.0              0.0        0.478745      0.021766      0  
2              1.0              0.0        0.452598      0.012449      1  
3              1.0              1.0        0.034076      0.212235      0  
4              0.0              1.0        0.058030      0.017465      1  

[5 rows x 46 columns]

In [7]:
# Selecting the Feature, by remove the unused feature
feature = ['Churn']
train_feature = df.drop(feature, axis=1)

# Set The Target
train_target = df["Churn"]

In [8]:
# Show the Feature
train_feature.head(5)

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  Partner_1  \
0       0.0       1.0              0.0              1.0        0.0        1.0   
1       1.0       0.0              0.0              1.0        1.0        0.0   
2       1.0       0.0              0.0              1.0        1.0        0.0   
3       1.0       0.0              0.0              1.0        1.0        0.0   
4       0.0       1.0              0.0              1.0        1.0        0.0   

   Dependents_0  Dependents_1    tenure  PhoneService_0  ...  Contract_0  \
0           0.0           1.0  0.013889             0.0  ...         0.0   
1           0.0           1.0  0.472222             1.0  ...         0.0   
2           0.0           1.0  0.027778             1.0  ...         0.0   
3           0.0           1.0  0.625000             0.0  ...         0.0   
4           0.0           1.0  0.027778             1.0  ...         0.0   

   Contract_1  Contract_2  PaperlessBilling_0  PaperlessBilling_1  \
0         0.0         1.0                 0.0                 1.0   
1         1.0         0.0                 1.0                 0.0   
2         0.0         1.0                 0.0                 1.0   
3         1.0         0.0                 1.0                 0.0   
4         0.0         1.0                 0.0                 1.0   

   PaymentMethod_0  PaymentMethod_1  PaymentMethod_2  MonthlyCharges  \
0              0.0              0.0              1.0        0.250169   
1              0.0              1.0              0.0        0.478745   
2              0.0              1.0              0.0        0.452598   
3              0.0              1.0              1.0        0.034076   
4              0.0              0.0              1.0        0.058030   

   TotalCharges  
0      0.003420  
1      0.021766  
2      0.012449  
3      0.212235  
4      0.017465  

[5 rows x 45 columns]

In [27]:
# Split Data
from sklearn.model_selection import train_test_split, cross_val_score
X_train, X_test, y_train, y_test = train_test_split(train_feature ,train_target, shuffle = True, test_size=0.9, random_state=1)

In [28]:
# Show the training data
X_train.head()

gender_0  gender_1  SeniorCitizen_0  SeniorCitizen_1  Partner_0  \
4342       0.0       1.0              1.0              0.0        1.0   
775        0.0       1.0              0.0              1.0        0.0   
4871       1.0       0.0              0.0              1.0        0.0   
4648       1.0       0.0              0.0              1.0        1.0   
5209       1.0       0.0              0.0              1.0        0.0   

      Partner_1  Dependents_0  Dependents_1    tenure  PhoneService_0  ...  \
4342        0.0           0.0           1.0  0.722222             1.0  ...   
775         1.0           0.0           1.0  1.000000             1.0  ...   
4871        1.0           1.0           0.0  0.805556             1.0  ...   
4648        0.0           0.0           1.0  0.027778             1.0  ...   
5209        1.0           0.0           1.0  1.000000             1.0  ...   

      Contract_0  Contract_1  Contract_2  PaperlessBilling_0  \
4342         0.0         1.0         1.0                 1.0   
775          0.0         1.0         1.0                 0.0   
4871         0.0         1.0         0.0                 0.0   
4648         0.0         0.0         1.0                 0.0   
5209         0.0         1.0         1.0                 0.0   

      PaperlessBilling_1  PaymentMethod_0  PaymentMethod_1  PaymentMethod_2  \
4342                 0.0              0.0              1.0              1.0   
775                  1.0              1.0              0.0              0.0   
4871                 1.0              1.0              0.0              0.0   
4648                 1.0              0.0              0.0              1.0   
5209                 1.0              1.0              0.0              0.0   

      MonthlyCharges  TotalCharges  
4342        0.503205      0.357814  
775         0.931680      0.930030  
4871        0.061825      0.049571  
4648        0.057692      0.000144  
5209        0.925776      0.090524  

[5 rows x 45 columns]

In [29]:
# Import Library
from sklearn.neural_network import MLPClassifier

# Fitting Model
mlp = MLPClassifier(hidden_layer_sizes=(5), activation = 'relu', solver = 'adam',max_iter= 10000, verbose = True)
mlp = mlp.fit(X_train,y_train)

# Prediction to Test Dataset
y_predmlp = mlp.predict(X_test)

Iteration 1, loss = 0.88623568
Iteration 2, loss = 0.84681097
Iteration 3, loss = 0.81219798
Iteration 4, loss = 0.78151440
Iteration 5, loss = 0.75434980
Iteration 6, loss = 0.73148409
Iteration 7, loss = 0.71218122
Iteration 8, loss = 0.69444846
Iteration 9, loss = 0.68050532
Iteration 10, loss = 0.66775739
Iteration 11, loss = 0.65775134
Iteration 12, loss = 0.64860027
Iteration 13, loss = 0.64047930
Iteration 14, loss = 0.63345844
Iteration 15, loss = 0.62760996
Iteration 16, loss = 0.62236047
Iteration 17, loss = 0.61748793
Iteration 18, loss = 0.61315049
Iteration 19, loss = 0.60923764
Iteration 20, loss = 0.60529627
Iteration 21, loss = 0.60200470
Iteration 22, loss = 0.59857817
Iteration 23, loss = 0.59546484
Iteration 24, loss = 0.59250588
Iteration 25, loss = 0.58956997
Iteration 26, loss = 0.58702377
Iteration 27, loss = 0.58429914
Iteration 28, loss = 0.58167377
Iteration 29, loss = 0.57918636
Iteration 30, loss = 0.57670996
Iteration 31, loss = 0.57442057
Iteration 32, los

In [31]:
print('Number of Layer =', mlp.n_layers_)
print('Number of Iteration =', mlp.n_iter_)
print('Current loss computed with the loss function =', mlp.loss_)

Number of Layer = 3
Number of Iteration = 1021
Current loss computed with the loss function = 0.3815079867518293


In [25]:
# Import the metrics class
from sklearn import metrics


# Confussion Matrix
cnf_matrixmlp = metrics.confusion_matrix(y_test, y_predmlp)
cnf_matrixmlp

array([[1397,  188],
       [ 225,  303]])

In [32]:
# Show the Accuracy, Precision, Recall, F1, etc.
acc_mlp = metrics.accuracy_score(y_test, y_predmlp)
prec_mlp = metrics.precision_score(y_test, y_predmlp)
rec_mlp = metrics.recall_score(y_test, y_predmlp)
f1_mlp = metrics.f1_score(y_test, y_predmlp)
kappa_mlp = metrics.cohen_kappa_score(y_test, y_predmlp)

print("Accuracy:", acc_mlp)
print("Precision:", prec_mlp)
print("Recall:", rec_mlp)
print("F1 Score:", f1_mlp)
print("Cohens Kappa Score:", kappa_mlp)

Accuracy: 0.7854551191039596
Precision: 0.5981970379909852
Recall: 0.557957957957958
F1 Score: 0.5773772529521443
Cohens Kappa Score: 0.43384732743817533


## Q1

**Split Data:** Splitting the dataset into two parts: the training set and the test set.

- `X_train` and `X_tes`t will contain the features (independent variables) for the training and testing datasets, respectively.
- `y_train` and `y_test` will contain the corresponding target labels (the 'Churn' column in this case) for the training and testing datasets, respectively.
- `shuffle=True` ensures that the data is randomly shuffled before splitting.
- `test_size=0.3` means that 30% of the data will be used for testing, and 70% for training.
- `random_state=1` sets a seed for the random number generator to ensure reproducibility. Different values will result in different random splits

**Confussion Matrix:** A table that is often used to describe the performance of a classification model. It shows the counts of True Positives (TP), True Negatives (TN), False Positives (FP), and False Negatives (FN).

- `y_test` contains the true labels of the test dataset
- `y_predmlp` contains the predicated labels generated by the MLPClassifier

**Other definitions:**
- `accuracy_score` calculates the accuracy of the model, which is the proportion of correctly predicted instances.
- `precision_score` calculates precision, which is the ratio of true positives to the total predicted positives. It measures how many of the predicted positive cases were correct.
- `recall_score` calculates recall, which is the ratio of true positives to the total actual positives. It measures how many of the actual positive cases were correctly predicted.
- `f1_score` calculates the F1 score, which is the harmonic mean of precision and recall. It provides a balance between precision and recall.
- `cohen_kappa_score` calculates Cohen's Kappa Score, which measures the agreement between the predicted and actual labels while taking into account the possibility of agreement by chance.

## Q2

**Training 70%:**

- Number of Layer = 3
- Number of Iteration = 184
- Current loss computed with the loss function = 0.41783620905746904
- Accuracy: 0.8045433033601515
- Precision: 0.6171079429735234
- Recall: 0.5738636363636364
- F1 Score: 0.5947006869479882
- Cohens Kappa Score: 0.46614353767256744

**Training 10%:**

- Number of Layer = 3
- Number of Iteration = 1021
- Current loss computed with the loss function = 0.3815079867518293
- Accuracy: 0.7854551191039596
- Precision: 0.5981970379909852
- Recall: 0.557957957957958
- F1 Score: 0.5773772529521443
- Cohens Kappa Score: 0.43384732743817533


Main observations:

1. Number of Iteration increased on the 10%, this is because training on smaller dataset can make convergence slower.
2. Accuracy, Precision, Recall, F1 and CKS a bit better on the larger training dataset.

The deference between both datasets are too small (~0.02), hence, I think training with 10% is acceptable to save computational resources.

In [35]:
# Import libraries
import pandas as pd
from sklearn.datasets import make_regression
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.multioutput import MultiOutputRegressor
from sklearn.metrics import mean_squared_error

# Load the dataset with TotalCharges as the target variable
url = 'https://raw.githubusercontent.com/rc-dbe/bigdatacertification/master/dataset/churn_trasnsformed_new.csv'
df = pd.read_csv(url, sep=',')

# Remove "Unnamed: 0" column
df = df.drop("Unnamed: 0", axis=1)

# initialize min-max scaler
mm_scaler = MinMaxScaler()
column_names = df.columns.tolist()

# Transform all attributes
df[column_names] = mm_scaler.fit_transform(df[column_names])
df.sort_index(inplace=True)

# Set TotalCharges as the target variable
target = 'TotalCharges'
feature_cols = [col for col in df.columns if col != target]

train_feature = df[feature_cols]
train_target = df[[target]]

# Split the data into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(train_feature, train_target, test_size=0.3, random_state=1)

# Train a Linear Regression model
regressor = GradientBoostingRegressor()
model = MultiOutputRegressor(regressor)
model.fit(x_train, y_train)

# Prediction to Test Dataset
y_predict = model.predict(x_test)

# Calculate the mean squared error (MSE)
mse = mean_squared_error(y_test, y_predict)
print("Mean Squared Error (MSE):", mse)

Mean Squared Error (MSE): 0.030313632383959645


## Q3

The MSE deference between training the model with 70% of the dataset is 0.030323568041804472 while it is 0.03331955320512922 on the 10%

The deference is 0.002995985163325 (~0.003) which I think is negligible.